MINOR: Start using Response and replace IOException in EmbeddedConnectCluster for failures

kkonstantine opened the pull request:
Inspecting the Response status code and entity is useful under testing. Therefore Connect's integration tests will benefit to transition from handling IOExceptions to inspecting the status code of the REST calls. 

Furthermore, catching a non-retriable IOException and wrapping it within ConnectException simplifies how lambdas are written for tests as well how failed REST responses are handled.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kkonstantine committed:
MINOR: Start using Response and replace IOException in EmbeddedConnectCluster for failures

kkonstantine commented:
@rhauch this is a PR that hopefully can be backported quite far back and can be used to improve integration testing with Connect. 

Let me know what you think and if it this looks good, whether you need a separate PR for older branches.

kkonstantine committed:
MINOR: Fix typos

kkonstantine committed:
MINOR: Small fixes including javadocs

rhauch commented:
ok to test

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -263,36 +266,37 @@ public void startConnect() {
      * @throws ConnectRestException if REST api returns error status
      */
-    public void configureConnector(String connName, Map<String, String> connConfig) throws IOException {
+    public String configureConnector(String connName, Map<String, String> connConfig) {
```
seems the return value is never used?

rhauch replied:
Isn't a string returned on line 283?

rhauch replied:
BTW, I think changing the return type from `void` to something else would be considered backward compatible, so this type of change is acceptable.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -263,36 +266,37 @@ public void startConnect() {
      * @param connConfig the intended configuration
-     * @throws IOException          if call to the REST api fails.
+     * @throws ConnectException if the configuration fails to be serialized
      * @throws ConnectRestException if REST api returns error status
```
Don't we also need to change the description to something like:
```suggestion
     * @throws ConnectRestException if REST api returns error status or the REST API fails
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -263,36 +266,37 @@ public void startConnect() {
      * @param connName   the name of the connector
      * @param connConfig the intended configuration
-     * @throws IOException          if call to the REST api fails.
+     * @throws ConnectException if the configuration fails to be serialized
```
```suggestion
     * @throws ConnectException if the configuration fails to be serialized or if the request could not be sent
```

kkonstantine replied:
Added with the other changes

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -263,36 +266,37 @@ public void startConnect() {
         }
-        if (status >= HttpServletResponse.SC_BAD_REQUEST) {
-            throw new ConnectRestException(status, "Could not execute PUT request");
+        Response response = executePut(url, content);
```
The `executePut` can actually throw `ConnectException`, so we should reflect that in the JavaDoc

kkonstantine replied:
Good point. Added

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -305,50 +309,63 @@ public void deleteConnector(String connName) throws IOException {
     }
 
-    public String adminEndpoint(String resource) throws IOException {
+    public String adminEndpoint(String resource) {
```
Since the `IOException` was declared on the signature, it's clear that calling code has to handle this case. But because `ConnectException` is a runtime exception, we should probably add JavaDoc for this method to explain that `ConnectException` will be thrown if the endpoint is not known.

kkonstantine replied:
Missed that these `public` methods were missing a javadoc. Added.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -305,50 +309,63 @@ public void deleteConnector(String connName) throws IOException {
     }
 
-    public String endpointForResource(String resource) throws IOException {
+    public String endpointForResource(String resource) {
```
Since the `IOException` was declared on the signature, it's clear that calling code has to handle this case. But because `ConnectException` is a runtime exception, we should probably add JavaDoc for this method to explain that `ConnectException` will be thrown if the endpoint is not known.

kkonstantine replied:
Same as above

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -363,87 +380,97 @@ public EmbeddedKafkaCluster kafka() {
+     * @return the response to the GET request
+     * @throws ConnectException if execution of the GET request fails
+     */
+    public Response executeGet(String url) {
```
Changing the return type from `int` to `Response` would be a breaking change. While this test harness is technically not part of the public API of Connect, we did intend that other projects could use this harness as long as they understand that we might change the API in breaking ways. 

But IMO we should still _try_ to evolve the API in ways that are backward compatible, and so I would suggest adding new methods that return `Response`, having the existing methods delegate to them, and deprecate the old methods.

kkonstantine replied:
That's an excellent point. I overlooked this possibility. 

I've deprecated all the `execute*` methods. Since there can't be overloading on the return type alone I've added the new methods as `request*`. Hope that works.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/SessionedProtocolIntegrationTest.java:
```diff
@@ -162,7 +162,7 @@ public void ensureInternalEndpointIsSecured() throws Throwable {
         assertEquals(
             FORBIDDEN.getStatusCode(),
-            connect.executePost(connectorTasksEndpoint, "[]", invalidSignatureHeaders)
+            connect.executePost(connectorTasksEndpoint, "[]", invalidSignatureHeaders).getStatus()
```
Technically, we shouldn't *have* to change this test if we were to maintain backward compatibility as I suggested on a different comment. But if we were to deprecate the existing `executePost` method that returns an int (and similar methods), I think it's worth changing this test to use the new methods that do return the `Response` to prevent a deprecation warning and to show users an example of how to migrate to the new methods.

kkonstantine replied:
I agree. Using the new methods now.

rhauch reviewed (commented):
@kkonstantine, thanks for the proposal. And apologies for sending all my comments as individual -- not sure what I was thinking.

As mentioned in those comments, I think we should try to make this change in a backward compatible manner just in case there are other projects that are using these methods and expecting the `int` return types. And since that would be relatively simple, IMO it's worth doing. WDYT?

Otherwise, very nice PR.

kkonstantine committed:
MINOR: Deprecate existing methods and add missing javadocs

kkonstantine reviewed (commented):
Thanks @rhauch 
I appreciate the review! Good point w/r/t deprecation. 
I believe I've addressed your points, feel free to take another look when you can.

kkonstantine commented:
After fixing a few minor checkstyle issues, this seems ready now.

kkonstantine committed:
MINOR: Introduce common assertions under EmbeddedConnectClusterAssertions

rhauch commented:
ok to test

kkonstantine commented:
@rhauch I added a small fix to the deprecated method `executeGet` to reinstate the correct return type (it was `String` not `int`) and rebased this PR. Based on some testing I did, the old methods maintain compatibility. They will break during upgrade only if a repo is failing on the use of deprecated methods. 

Lmk how this looks now.

kkonstantine commented:
FYI, last build succeeded for JDK 8 and Scala 2.12
and failed with an irrelevant test 
(`kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`) for JDK 11 and Scala 2.13.

rhauch commented:
retest this please

rhauch commented:
ok to test

rhauch reviewed (approved):
LGTM. Thanks for the improvements, @kkonstantine!
